File: backend_loyalty/backend_loyalty/models/review.js


const { DataTypes } = require('sequelize');
const sequelize = require('../config/database');
const User = require('./user'); 
const Product = require('./product'); 

const Review = sequelize.define('Review', {
  id: {
    type: DataTypes.BIGINT.UNSIGNED,
    allowNull: false,
    primaryKey: true,
    autoIncrement: true,
  },
  rating: {
    type: DataTypes.DOUBLE,
    allowNull: true,
  },
  comment: {
    type: DataTypes.STRING,
    allowNull: true,
  },
  user_id: {
    type: DataTypes.BIGINT.UNSIGNED,
    
  },
  product_id: {
    type: DataTypes.BIGINT.UNSIGNED,
    
  },
  created_at: {
    type: DataTypes.DATE,
    allowNull: true,
  },
  updated_at: {
    type: DataTypes.DATE,
    allowNull: true,
  },
  // deletedAt: {
  //   type: DataTypes.DATE,
  //   allowNull: true,
  // }
  },{
  timestamps: false, 
  paranoid: true,
  tableName: 'reviews',
  
  });

Review.belongsTo(User, { foreignKey: 'user_id', as: 'users',key: 'id', onDelete: 'CASCADE' });
Review.belongsTo(Product, { foreignKey: 'product_id', as: 'products',key: 'id', onDelete: 'CASCADE'  });

module.exports = Review;

File: Loyalty-App/loyalty_app/lib/model/review.dart

import 'package:loyalty_app/model/product.dart';
import 'package:loyalty_app/model/user.dart';

class Review {
  
int? id;
  double rating;
  int productId;
  int ?userId;
  String comment;
  Product? product;
  User?user;

  Review({
 this.id,
    required this.rating,
    required this.productId,
    this.userId,
    required this.comment,
    this.product,
    this.user,
  });

  factory Review.fromJson(Map<String, dynamic> json) {
    return Review(
      id: json['id'] as int?,
      rating: (json["rating"] as num?)?.toDouble() ?? 0.0,
      productId: json["product_id"] as int? ?? 0,
      userId: json["user_id"]as int? ?? 0,
      comment: json["comment"] as String? ?? "",
       user: json['users'] != null ? User.fromJson(json['users'] as Map<String, dynamic>) : null,
    
      product: json['product'] != null ? Product.fromJson(json['product'] as Map<String, dynamic>) : null,
    );
  }
  Map<String, dynamic> toJson() => {
        
        "id": id,
        'rating': rating,
        'product_id': productId,
        'comment': comment,
        "product": product?.toJson(),
        "user_id":userId,
        "user":user?.toJson(),
       
       
      };
}
